server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ObjectService.scala

```diff
@@ -144,21 +147,31 @@ class ObjectService(
           result <- metadata match {
             case None =>
               Kleisli.pure[IO, TraceId, MetadataResponse](MetadataResponse.RemoteNotFound(context.storageLink))
-            case Some(AdaptedGcsMetadata(lastLockedBy, crc32c, generation)) =>
+            case Some(AdaptedGcsMetadata(lock, crc32c, generation)) =>
               val localAbsolutePath = config.workingDirectory.resolve(req.localObjectPath.asPath)
               val res = for {
                 calculatedCrc32c <- Crc32c.calculateCrc32ForFile(localAbsolutePath, blockingEc)
                 syncStatus <- if (calculatedCrc32c == crc32c) IO.pure(SyncStatus.Live)
                 else {
                   for {
-                    cachedGeneration <- metadataCache.get.map(_.get(req.localObjectPath.asPath))
-                    status <- cachedGeneration match {
-                      case Some(cachedGen) => if (cachedGen.generation == generation) IO.pure(SyncStatus.LocalChanged) else IO.pure(SyncStatus.RemoteChanged)
-                      case None => IO.pure(SyncStatus.Desynchronized)
+                    metaOpt <- metadataCache.get.map(_.get(req.localObjectPath))
+                    status <- metaOpt match {
+                      case Some(meta) =>
+                        meta.localFileStateInGCS match { //TODO: shall we check the file has been localized before calculating crc32c
+                          case Some(previousFileState) =>
+                            if (previousFileState.generation == generation)
+                              IO.pure(SyncStatus.LocalChanged) <* logger.ctxWarn[PostContext, String]("local file has changed, but it hasn't been delocalized yet").run(loggingContext)
+                            else IO.pure(SyncStatus.RemoteChanged) <* logger.ctxWarn[PostContext, String]("remote has changed").run(loggingContext)
+                          case None =>
+                            IO.pure(SyncStatus.Desynchronized) <* logger.ctxError[PostContext, String]("We don't find local generation for a localized file. This shouldn't happen").run(loggingContext)
+                        }
+                      case None =>
+                        IO.pure(SyncStatus.Desynchronized) <* logger.ctxError[PostContext, String]("We don't find local cache for a localized file. This shouldn't happen").run(loggingContext) //TODO: this shouldn't be possible because we should have the file in cache if it has been localized
```

Comment:
There are many ways this could happen, e.g. transient failures, direct file system manipulation on the VM (`cp` from terminal, `!touch` from notebook, etc). Ideally Welder should handle this scenario gracefully if it wants to be more robust. This goes back to the metadata caching discussion - I would change the current behavior as follows:

- metadata cache always tracks the latest known GCS metadata
- /checkMetadata doesn't check the cache, it always reads from GCS and then updates the cache accordingly
- Everything else can reference the cache for reading metadata values (optionally forcing a GET if we've never cached the value yet)
- Don't touch the cache in /localize

Reply:
If we don't update cache's generation when localizing, how do we know the generation before user modifies the file? That generation is used for delocalizing and deleting

Reply:
Here is what I would propose:

```
safeDelocalize:
  meta = metaCache.get(path, fetchIfMissing=true)
```

It wouldn't hurt to cache on localization, but I wouldn't want to depend on that. I think it's a risky approach to try and control all mechanisms by which a file could get onto the VM.

If we did that, we'd also need to slightly change semantics (as we discussed previously: https://github.com/DataBiosphere/leonardo/blob/develop/docker/jupyter/custom/jupyter_delocalize.py#L217) - Jupyter should not update the file until after Welder. Then `/safeDelocalize` could compare the current crc32cs before deciding whether to update.

Reply:
@Qi77Qi convinced me in Slack. There are downsides to both approaches but the current way seems likely to behave better during standard usage.